### PR TITLE
Update `LightningDataModule` docs

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Part of docs revamp. Updating `LightiningDataModule` docs to latest.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: docs/source/extensions/datamodules.rst [hunk 1]
```diff
@@ -264,11 +268,11 @@ Use this method to generate the val dataloader. Usually you just wrap the datase
             return DataLoader(self.mnist_val, batch_size=64)
 
 
-.. _datamodule-test-dataloader-label:
-
 test_dataloader
```
Comment (rohitgr7) [inline docs/source/extensions/datamodules.rst:271]:
```suggestion
.. _datamodule-test-dataloader-label:

test_dataloader
```

File: docs/source/extensions/datamodules.rst [hunk 2]
```diff
@@ -195,16 +196,17 @@ settings.
             MNIST(os.getcwd(), train=False, download=True, transform=transforms.ToTensor())
 
 
-.. warning:: ``prepare_data`` is called from a single process (e.g. GPU 0). Do not use it to assign state (`self.x = y`).
+.. warning:: ``prepare_data`` is called from a single process across all nodes. It is not recommended to assign state here (e.g. ``self.x = y``).
```
Comment (rohitgr7) [inline docs/source/extensions/datamodules.rst:199]: can someone confirm this?

Comment (carmocca) [reply docs/source/extensions/datamodules.rst:199]: Depends on the value of `LightningModule.prepare_data_per_node`:

https://github.com/PyTorchLightning/pytorch-lightning/blob/2036dfb5df2a48056d1c043ee311aed793187070/pytorch_lightning/trainer/connectors/data_connector.py#L159-L160

**Review by carmocca — APPROVED** [review]

**Comment by rohitgr7** [comment]
You can review the docs here: https://139808-178626720-gh.circle-artifacts.com/0/html/extensions/datamodules.html

File: docs/source/extensions/datamodules.rst [hunk 3]
```diff
@@ -220,25 +222,25 @@ There are also data operations you might want to perform on every GPU. Use setup
...
-.. note:: ``setup`` is called from every process. Setting state here is okay.
-.. note:: ``teardown`` can be used to clean up the state. It is also called from every process
+.. note:: :ref:`setup<extensions/datamodules:setup>` is called from every process across all the nodes. Setting state here is recommended.
+.. note:: :ref:`teardown<extensions/datamodules:teardown>` can be used to clean up the state. It is also called from every process across all the nodes.
 
 
 train_dataloader
 ^^^^^^^^^^^^^^^^
-Use this method to generate the train dataloader. Usually you just wrap the dataset you defined in ``setup``.
+Use the :meth:`~pytorch_lightning.core.hooks.DataHooks.train_dataloader` method to generate the train dataloader(s).
```
Comment (awaelchli) [inline docs/source/extensions/datamodules.rst:241]:
```suggestion
Use the :meth:`~pytorch_lightning.core.hooks.DataHooks.train_dataloader` method to generate the training dataloader(s).
```

File: docs/source/extensions/datamodules.rst [hunk 4]
```diff
@@ -252,7 +254,9 @@ Use this method to generate the train dataloader. Usually you just wrap the data
 
 val_dataloader
 ^^^^^^^^^^^^^^
-Use this method to generate the val dataloader. Usually you just wrap the dataset you defined in ``setup``.
+Use the :meth:`~pytorch_lightning.core.hooks.DataHooks.val_dataloader` method to generate the val dataloader(s).
```
Comment (awaelchli) [inline docs/source/extensions/datamodules.rst:257]:
```suggestion
Use the :meth:`~pytorch_lightning.core.hooks.DataHooks.val_dataloader` method to generate the validation dataloader(s).
```

File: docs/source/extensions/datamodules.rst [hunk 5]
```diff
@@ -282,7 +288,8 @@ Use this method to generate the test dataloader. Usually you just wrap the datas
 
 predict_dataloader
 ^^^^^^^^^^^^^^^^^^
-Returns a special dataloader for inference. This is the dataloader that the Trainer
+Use the :meth:`~pytorch_lightning.core.hooks.DataHooks.predict_dataloader` method to generate the predict dataloader(s).
```
Comment (awaelchli) [inline docs/source/extensions/datamodules.rst:291]:
```suggestion
Use the :meth:`~pytorch_lightning.core.hooks.DataHooks.predict_dataloader` method to generate the prediction dataloader(s).
```

File: docs/source/extensions/datamodules.rst [hunk 6]
```diff
@@ -292,76 +299,68 @@ Returns a special dataloader for inference. This is the dataloader that the Trai
...
-        def __init__(self, train_transforms, val_transforms, test_transforms):
-            super().__init__()
-            self.train_transforms = train_transforms
-            self.val_transforms = val_transforms
-            self.test_transforms = test_transforms
+.. automethod:: pytorch_lightning.core.datamodule.LightningDataModule.on_predict_dataloader
+    :noindex:
+
+teardown
+~~~~~~~~~~~~~~~~~~~~~
```
Comment (awaelchli) [inline docs/source/extensions/datamodules.rst:360]: too long underline

File: docs/source/extensions/datamodules.rst [hunk 7]
```diff
@@ -292,76 +299,68 @@ Returns a special dataloader for inference. This is the dataloader that the Trai
...
-            return batch
+on_after_batch_transfer
+^^^^^^^^^^^^^^^^^^^^^^^
 
+.. automethod:: pytorch_lightning.core.datamodule.LightningDataModule.on_after_batch_transfer
+    :noindex:
 
-.. note:: This hook only runs on single GPU training and DDP (no data-parallel).
+on_load_checkpoint
+~~~~~~~~~~~~~~~~~~
```
Comment (awaelchli) [inline docs/source/extensions/datamodules.rst:325]: from here on, the title levels are inconsistent, causing the index to be confused:

image

Not sure why.

File: docs/source/extensions/datamodules.rst [hunk 8]
```diff
@@ -292,76 +299,68 @@ Returns a special dataloader for inference. This is the dataloader that the Trai
 
     class MNISTDataModule(pl.LightningDataModule):
         def predict_dataloader(self):
-            return DataLoader(self.mnist_test, batch_size=64)
+            return DataLoader(self.mnist_predict, batch_size=64)
 
 
 transfer_batch_to_device
 ^^^^^^^^^^^^^^^^^^^^^^^^
-Override to define how you want to move an arbitrary batch to a device.
```
Comment (awaelchli) [inline docs/source/extensions/datamodules.rst:300]: I preferred the previous way of separating API docs from user guide.

**Review by awaelchli — APPROVED** [review]

File: docs/source/extensions/datamodules.rst (hunk 8 above)
Comment (rohitgr7) [reply docs/source/extensions/datamodules.rst:300]: won't this one be better? user will get the API arguments + the example was just repeated here and already present in the API docs.

File: docs/source/extensions/datamodules.rst (hunk 7 above)
Comment (rohitgr7) [reply docs/source/extensions/datamodules.rst:325]: fixed!

File: docs/source/extensions/datamodules.rst [hunk 9]
```diff
@@ -162,25 +161,27 @@ Here's a more realistic, complex DataModule that shows how much more reusable th
...
+- :ref:`test_dataloader<extensions/datamodules:test_dataloader>`
+- :ref:`predict_dataloader<extensions/datamodules:predict_dataloader>`
 
 
 prepare_data
-^^^^^^^^^^^^
-Use this method to do things that might write to disk or that need to be done only from a single process in distributed
-settings.
+~~~~~~~~~~~~
+Use :meth:`~pytorch_lightning.core.hooks.DataHooks.prepare_data` method to do things that might write to disk or that need to be done only from a single process in distributed
```
Comment (tchaton) [inline docs/source/extensions/datamodules.rst:183]:
```suggestion
Downloading and saving data with multiple processes (distributed settings) will result in corrupted data.  Lightning ensures the :meth:`~pytorch_lightning.core.hooks.DataHooks.prepare_data` is called only within a single process, so you can safely add your downloading logic within.
```

**Review by tchaton — APPROVED** [review]
LGTM !
